src/server/tiering/external_alloc.cc

```diff
@@ -478,12 +496,23 @@ void ExternalAllocator::FreePage(Page* page, SegmentDescr* owner, size_t block_s
   BinIdx bidx = ToBinIdx(block_size);
 
   // Remove fast allocation reference.
-  if (free_pages_[bidx] == page) {
-    free_pages_[bidx] = &empty_page;
+  if (page->free_chained) {
+    if (free_pages_[bidx] == page) {
+      free_pages_[bidx] = &empty_page;
```

Comment:
should not you set `page->next_free` in this case?

Reply:
Good catch 😳